lib/include/platform/atomic.h

```diff
@@ -0,0 +1,294 @@
+    #else
+        return Atomic_CompareAndSwap_u32_User_Override( pDestination, ulExchange, ulComparand );
+    #endif /* COMPILER_OPTION_GCC_ATOMIC_BUILTIN */
+}
+
+/**
+ * Atomic swap (pointers)
+ *
+ * @brief Atomically sets the address pointed to by *ppDestination to the value of *pExchange.
+ *
+ * @param[in, out] ppDestination  Pointer to memory location from where a pointer value is to be loaded and writen back to.
+ * @param[in] pExchange           Pointer value to be written to *ppDestination.
+ *
+ * @return The initial value of *ppDestination.
+ */
+static FORCE_INLINE_FLATTEN void * Atomic_SwapPointers_p32( void * volatile * ppDestination,
+                                                            void * pExchange )
+{
+    #if ( COMPILER_OPTION_GCC_ATOMIC_BUILTIN == 1 )
+        void * pReturnValue;
+        __atomic_exchange( ppDestination, &pExchange, &pReturnValue, __ATOMIC_SEQ_CST );
+
+        return pReturnValue;
+    #else
+        return Atomic_SwapPointers_p32_User_Override( ppDestination, pExchange );
+    #endif /* COMPILER_OPTION_GCC_ATOMIC_BUILTIN */
+}
+
+/**
+ * Atomic compare-and-swap (pointers)
+ *
+ * @brief Performs an atomic compare-and-swap operation on the specified pointer values.
+ *
+ * @param[in, out] ppDestination Pointer to memory location from where a pointer value is to be loaded and checked.
+ * @param[in] pExchange          If condition meets, write this value to memory.
+ * @param[in] pComparand         Swap condition, checks and waits for *ppDestination to be equal to *pComparand.
+ *
+ * @return bool True for swapped, false for not swapped.
+ *
+ * @note This function only swaps *ppDestination with pExchange, if previous *ppDestination value equals pComparand.
+ */
+static FORCE_INLINE_FLATTEN bool Atomic_CompareAndSwapPointers_p32( void * volatile * ppDestination,
+                                                                    void * pExchange,
+                                                                    void * pComparand )
+{
+    #if ( COMPILER_OPTION_GCC_ATOMIC_BUILTIN == 1 )
+        return __atomic_compare_exchange( ppDestination, &pComparand, &pExchange, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST );
+    #else
+        return Atomic_CompareAndSwapPointers_p32_User_Override( ppDestination, pExchange, pComparand );
+    #endif /* COMPILER_OPTION_GCC_ATOMIC_BUILTIN */
+}
+
+
+/*----------------------------- Arithmetic ------------------------------*/
+
+/**
+ * Atomic add
+ *
+ * @brief Adds count to the value of the specified 32-bit variable as an atomic operation.
+ *
+ * @param[in,out] pAddend  Pointer to memory location from where value is to be loaded and written back to.
+ * @param[in] lCount       Value to be added to *pAddend.
+ *
+ * @return previous *pAddend value.
+ */
+static FORCE_INLINE_FLATTEN int32_t Atomic_Add_i32( int32_t volatile * pAddend,
```

Comment:
Here we need to complete our evaluation of signed- vs. unsigned- integers...

Reply:
updated per offline discussion.